### PR TITLE
[TA1646] Fixing memory leak in replication module

### PR DESCRIPTION
**Pull request by mynktl**
Changes : 
Memory leak fix for replication module reported by Valgrind. Test was performed on `istgt_integration_test`.

Details of memory leak before this changes:
https://rally1.rallydev.com/slm/attachment/238766730428/valgrind_output_before_changes
Details of memory leak with this changes:
https://rally1.rallydev.com/slm/attachment/238767578200/valgrind_output_after_memory_leak_fix

### CODE REVIEW

File: src/data_conn.c [hunk 1]
```diff
@@ -259,6 +298,7 @@ handle_data_conn_error(replica_t *r)
 
 	if (r->iofd == -1) {
 		REPLICA_ERRLOG("repl %s %d %p\n", r->ip, r->port, r);
+		cleanup_dead_replica(r);
```
Comment (vishnuitta) [inline src/data_conn.c:301]: lets not worry about cleaning dead replicas..

File: src/data_conn.c [hunk 2]
```diff
@@ -335,6 +370,7 @@ handle_data_conn_error(replica_t *r)
 		pthread_cond_signal(&r->r_cond);
 	r->mgmt_eventfd2 = -1;
 	MTX_UNLOCK(&r->r_mtx);
+	destroy_mempool(&r->cmdq);
```
Comment (vishnuitta) [inline src/data_conn.c:373]: lets move this to free_replica and also check for any other left out allocations..

File: src/data_conn.c [hunk 3]
```diff
@@ -661,9 +697,6 @@ replica_thread(void *arg)
 		REPLICA_ERRLOG("epoll error for replica(%s:%d) err(%d)\n",
 		    r->ip, r->port, errno);
 initialize_error:
-		close(r->mgmt_eventfd2);
```
Comment (vishnuitta) [inline src/data_conn.c:664]: lets close r_mgmt_eventfd, and leave other fds to -1

File: src/replication.c [hunk 4]
```diff
@@ -1878,15 +1885,15 @@ inform_data_conn(replica_t *r)
...
+void
 free_replica(replica_t *r)
 {
 	pthread_mutex_destroy(&r->r_mtx);
 	pthread_cond_destroy(&r->r_cond);
 
-	free(r->mgmt_io_resp_hdr);
-	free(r->m_event1);
-	free(r->m_event2);
+	if (r->cmdq.ring)
```
Comment (vishnuitta) [inline src/replication.c:1894]: it will be already empty.. so, lets assert for empty cmdq

File: src/replication.c [hunk 5]
```diff
@@ -2363,8 +2370,29 @@ handle_mgmt_conn_error(replica_t *r, int sfd, struct epoll_event *events, int ev
 	}
 	MTX_UNLOCK(&r->spec->rq_mtx);
 
-	if (r->dont_free != 1)
+	if (r->mgmt_io_resp_hdr)
+		free(r->mgmt_io_resp_hdr);
```
Comment (vishnuitta) [inline src/replication.c:2374]: this code also may not be required if we are not cleaning up dead replicas..

File: src/data_conn.c (hunk 1 above)
Comment (mynktl) [reply src/data_conn.c:301]: ok

File: src/data_conn.c (hunk 2 above)
Comment (mynktl) [reply src/data_conn.c:373]: ok

File: src/data_conn.c (hunk 3 above)
Comment (mynktl) [reply src/data_conn.c:664]: ok

File: src/replication.c (hunk 4 above)
Comment (mynktl) [reply src/replication.c:1894]: ok

File: src/replication.c (hunk 5 above)
Comment (mynktl) [reply src/replication.c:2374]: ok

**Review by mynktl — COMMENTED** [review]
@vishnuitta I have addressed the review comments.
I have pushed the required changes from this PR to PR #54.

**Comment by mynktl** [comment]
Changes from this PR has been merged in PR #54. 
Changes not merged (to avoid complexity) :
- Destroying replica in data_conn.c (replica having `dont_free=1`)

**Pull request closed without merging**